proofs.go

```diff
@@ -1,93 +1,168 @@
 package smt
 
-import(
-    "hash"
-    "bytes"
-    "errors"
+import (
+	"bytes"
+	"errors"
+	"hash"
+	"math"
 )
 
+// SparseMerkleProof is a Merkle proof for a element in a SparseMerkleTree.
+type SparseMerkleProof struct {
+	// SideNodes is an array of the sibling nodes leading up to the leaf of the proof.
+	SideNodes [][]byte
+
+	// NonMembershipLeafData is the data of the unrelated leaf at the position
+	// of the key being proven, in the case of a non-membership proof.
+	NonMembershipLeafData []byte
+
+	// BitMask, in the case of a compact proof, is a bit mask of the sidenodes
+	// of the proof where an on-bit indicates that the sidenode at the bit's
+	// index is a placeholder. This is only set if the proof is compact.
+	BitMask []byte
+
+	// NumSideNodes, in the case of a compact proof, indicates the number of
+	// sidenodes in the proof when decompacted. This is only set if the proof is compact.
+	NumSideNodes int
```

Comment:
Might be worth to consider introducing a dedicated CompactSparseMerkleProof type that extends the `SparseMerkleProof` with these two fields?

Reply:
This was the first thing I considered, however the drawback is that in order to initialise a new CompatSparseMerkleProof instance, one would also have to construct and nest a SparseMerkleProof instance, which IMO makes the API too fiddly and complicated. The alternative would be to replicate the two fields, but it fields like having a struct with duplicated fields and sanity checks could be redundant.

Reply:
I don't fully understand how that would make the API fiddly and complicated (imho it could make it clearer). Let's not bother with this now. Happy to look into it after this gets merged.